src/editable.ts

```diff
@@ -0,0 +1,328 @@
+        ev.preventDefault();
+    });
+    container.addEventListener('mouseup', () => {
+        cursorCopy(container, false);
+        cursorMove(container, false);
+    });
+};
+
+const addDropzoneEvents = (context: Context): void => {
+    const { container } = context;
+    const dropzoneElems = container.querySelectorAll<SVGRectElement>('.dropzone');
+    dropzoneElems.forEach((dropzoneElem) => {
+        dropzoneElem.addEventListener('mouseup', (ev: MouseEvent) => handleDropzoneMouseUp(ev, context));
+    });
+};
+
+const addMouseEvents = (context: Context): void => {
+    const { container } = context;
+    const gateElems = getGateElems(container);
+    gateElems.forEach((gateElem) => {
+        gateElem.addEventListener('mousedown', (ev: MouseEvent) => handleGateMouseDown(ev, container));
+    });
+};
+
+// Event handlers
+const handleGateMouseDown = (ev: MouseEvent, container: HTMLElement): void => {
+    ev.stopPropagation();
+    _sourceTarget = ev.currentTarget as SVGGElement;
+
+    // Ctrl + Mousedown to copy. Mousedown only to move.
+    ev.ctrlKey ? cursorCopy(container, true) : cursorMove(container, true);
+};
+
+const handleDropzoneMouseUp = (ev: MouseEvent, context: Context): void | false => {
+    ev.stopPropagation();
+
+    const { container, operations, wires, renderFn } = context;
+
+    const currentTarget = ev.currentTarget as SVGGElement;
+
+    if (!currentTarget) return false;
+
+    const dataId = getDataId(currentTarget);
+    const parent = getParent(dataId, operations);
+    const index = splitDataId(dataId).pop();
+    const position = getDropzonePosition(currentTarget);
+
+    if (_sourceTarget == null) return false;
+
+    const sourceDataId = getDataId(_sourceTarget);
+    const sourceParent = getParent(sourceDataId, operations);
+    const sourceIndex = splitDataId(sourceDataId).pop();
+
+    if (index == null || sourceIndex == null) return false;
+
+    const newGate = getGate(sourceDataId, operations);
+    const wireY = getClosestWireY(ev.offsetY, wires);
+
+    // Not allow Measure gate to move vertically
+    if (wireY != null && newGate.gate !== 'measure') {
+        // wires[wireY] returns qubit name (i.e: 'q0')
+        // this remove 'q' and assign an index (i.e: 0)
+        const index = Number(wires[wireY].slice(1));
+        const [firstTarget, ...targetsExceptFirst] = newGate.targets;
+        // Reserve all other properties, only change qId
+        Object.assign(firstTarget, { ...firstTarget, qId: index });
+        // Reserve all other targets, only change first target
+        Object.assign(newGate, { ...newGate, targets: [firstTarget, ...targetsExceptFirst] });
+    }
+
+    // Remove source element if moving using Ctrl + Mousedown
+    if (!ev.ctrlKey) {
+        deleteAt(sourceParent, sourceIndex);
+    }
+
+    // If dropzone is left of gate, insert before gate.
+    // Otherwise, insert after.
+    if (position === 'left') {
+        insertBefore(parent, index, newGate);
```

Comment:
There's a logical error in here. If the gate being moved is coming from the right, this works. If the gate is being copied, this works. If the gate is being moved from the left, this results in that moved gate being placed to the right of the parent instead of the left. 

This is because the `index` value is set in L106, before the gate being moved is deleted at L134. If the gate being moved was originally to the right of parent, then there's no change and everything works. But if the gate was to the left, then parent is actually at `index - 1` now, so setting newGate to `index` puts it to the right of parent instead of to the left.

Reply:
This also impacts `insertAfter` below, where if newGate is being moved from the left of parent, then it'll end up two spaces to the right instead of immediately to the right.